File: apps/desktop/src/lib/telemetry.ts
import * as Sentry from "@sentry/react";
import { invoke } from "@tauri-apps/api/core";

/**
 * Crash reporting for the desktop and Android webview.
 *
 * The Rust side of this app reports through the Sentry Rust SDK; this covers the
 * half that runs in the webview, where a React render error or a broken IPC
 * bridge would otherwise only reach a `console.error` that, in a packaged app on
 * Windows or Android, nothing is attached to.
 *
 * Rust owns the decision. The DSN, the opt-out state and the install ID all come
 * from `telemetry_status`, rather than being baked in here separately — if the
 * webview made its own call, turning telemetry off would silence Rust while
 * JavaScript carried on reporting, which is worse than not offering the switch.
 */

declare const __APP_VERSION__: string;

interface TelemetryStatus {
  enabled: boolean;
  dsn: string | null;
  platform: string;
  install_id: string | null;
  notice_pending: boolean;
}

/**
 * Paths are the main way personal data leaks out of this app: an error like
 * "failed to load /home/alice/models/q4.gguf" carries the user's real name.
 * Mirrors `redact_sensitive` in hivebear-core, which is the tested copy — keep
 * the two in step.
 */
export function redact(input: string): string {
  return input
    .replace(/(\/home\/|\/Users\/)[^/\\]+/g, "$1<user>")
    .replace(/([A-Z]:\\Users\\)[^\\]+/gi, "$1<user>")
    .replace(/\b(sk-ant-|sk-|gsk_|hf_|xai-)[A-Za-z0-9_-]+/g, "[redacted-key]")
    .replace(/[\w.+-]+@[\w-]+\.[\w.-]+/g, "[email]");
}

let status: TelemetryStatus | null = null;

/** The resolved status, once `initTelemetry` has run. */
export function telemetryStatus(): TelemetryStatus | null {
  return status;
}

/**
 * Start reporting, if Rust says we should.
 *
 * Deliberately async and deliberately not awaited before the first render: it
 * costs one IPC round trip, and blocking paint on it would trade a real
 * user-visible delay for a marginal gain in coverage. Errors thrown in that
 * window are missed, which is the right way round — the alternative is
 * initialising before we know whether the user opted out.
 */
export async function initTelemetry(): Promise<void> {
  try {
    status = await invoke<TelemetryStatus>("telemetry_status");
  } catch {
    // If we cannot ask, do not report. Failing closed is the only safe default
    // for something the user is entitled to switch off.
    return;
  }

  if (!status.enabled || !status.dsn) return;

  Sentry.init({
    dsn: status.dsn,
    release: `hivebear@${__APP_VERSION__}`,
    environment: import.meta.env.DEV ? "development" : "production",

    tracesSampleRate: import.meta.env.DEV ? 1.0 : 0.1,

    // Never. This is someone else's computer.
    sendDefaultPii: false,

    beforeSend(event) {
      if (event.message) event.message = redact(event.message);
      for (const exception of event.exception?.values ?? []) {
        if (exception.value) exception.value = redact(exception.value);
      }
      for (const crumb of event.breadcrumbs ?? []) {
        if (crumb.message) crumb.message = redact(crumb.message);
      }
      return event;
    },
  });

  Sentry.setUser(status.install_id ? { id: status.install_id } : null);
  // The same bundle ships as desktop and as the Android webview; without this
  // the two are indistinguishable in one project.
  Sentry.setTag("platform", status.platform);
  Sentry.setTag("surface", "webview");
}

/** Report an error we are handling rather than rethrowing. */
export function reportError(
  error: unknown,
  context: Record<string, string> = {},
): void {
  Sentry.captureException(error, { tags: context });
}

File: apps/desktop/src-tauri/src/commands/telemetry.rs
use serde::Serialize;

use crate::telemetry;
use hivebear_core::telemetry::{resolve, TelemetryDecision};
use hivebear_core::Config;

/// What the webview needs to configure its own reporting.
///
/// The webview cannot read the compile-time DSN or the config file, and it must
/// not make its own decision about consent — otherwise turning telemetry off
/// would silence the Rust half of the app while the JavaScript half kept
/// reporting. Rust decides; the webview is told.
#[derive(Debug, Serialize)]
pub struct TelemetryStatus {
    /// Whether reporting is on. False for an opt-out *or* a source build.
    pub enabled: bool,
    /// The DSN to report to, or `None` when disabled. DSNs are not secret.
    pub dsn: Option<String>,
    /// `desktop` or `android` — the same web bundle ships as both.
    pub platform: String,
    /// The pseudonymous per-install ID, so both halves report as one user.
    pub install_id: Option<String>,
    /// Whether the first-run notice still needs showing.
    pub notice_pending: bool,
}

/// Report the resolved telemetry state to the webview.
#[tauri::command]
pub fn telemetry_status() -> TelemetryStatus {
    let config = Config::load();
    let decision = resolve(&config.telemetry);

    TelemetryStatus {
        enabled: decision.is_enabled(),
        dsn: match &decision {
            TelemetryDecision::Enabled { dsn } => Some(dsn.clone()),
            TelemetryDecision::Disabled(_) => None,
        },
        platform: telemetry::platform().to_string(),
        install_id: config.telemetry.install_id.clone(),
        notice_pending: decision.is_enabled() && !config.telemetry.notice_shown,
    }
}

/// Record that the first-run telemetry notice has been shown.
///
/// Opt-out reporting is only defensible if people are actually told, and this is
/// a GUI app with no console — so the notice lives in the UI and this is how the
/// UI says it has done its job.
#[tauri::command]
pub fn acknowledge_telemetry_notice() -> Result<(), String> {
    let mut config = Config::load();
    config.telemetry.notice_shown = true;
    config.save().map_err(|e| e.to_string())
}

// Turning reporting on and off deliberately does NOT have its own command. It
// is an ordinary config field, so it goes through `save_config` like every other
// setting — one write path, no chance of the two disagreeing about which is
// authoritative.
